src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb

```diff
@@ -3429,61 +3429,174 @@ class C
+                    Await Task.Delay(250)
 
-                Await state.WaitForAsynchronousOperationsAsync()
-                Await state.AssertNoCompletionSession()
-                Assert.Contains("System", state.GetLineTextFromCaretPosition())
+                    state.AssertNoCompletionSessionWithNoBlock()
+                    Assert.Contains("Sys", state.GetLineTextFromCaretPosition())
+                    Assert.DoesNotContain("System", state.GetLineTextFromCaretPosition())
+                    tcs.SetResult(True)
+
+                    Await state.WaitForAsynchronousOperationsAsync()
+                    Await state.AssertNoCompletionSession()
+                    Assert.Contains("System", state.GetLineTextFromCaretPosition())
+                Else
+                    Dim task1 As Task = Nothing
+                    Dim task2 As Task = Nothing
+
+                    Dim providerCalledHandler = Sub()
+                                                    task1 = Task.Run(Sub()
+                                                                         ' 2. Hang here as well.
+                                                                         Dim completionItem = state.GetSelectedItemOpt()
+                                                                     End Sub)
+                                                    Thread.Sleep(250)
+                                                    task2 = Task.Run(Sub()
+                                                                         Try
+                                                                             ' 3. Check that the other task is running/hanging.
+                                                                             Assert.Equal(TaskStatus.Running, task1.Status)
+                                                                             Assert.Contains("Sys", state.GetLineTextFromCaretPosition())
+                                                                             Assert.DoesNotContain("System", state.GetLineTextFromCaretPosition())
+                                                                         Finally
+                                                                             ' 4. Unblock the first task and the main thread.
+                                                                             tcs.SetResult(True)
+                                                                         End Try
+                                                                     End Sub)
+                                                End Sub
+
+                    Try
+                        AddHandler provider.ProviderCalled, providerCalledHandler
+
+                        ' In the new completion, when pressed <ctrl>-<space>, we have to wait for the aggregate operation to complete.
+                        ' 1. Hang here.
+                        state.SendCommitUniqueCompletionListItem()
```

Comment:
If this blocks until completion has been committed, aren't `task1` and `task2` already complete when it returns?  If so, what is the `WhenAll` for?

Reply:
`task1` can still been executing when `state.SendCommitUniqueCompletionListItem()` is done. I'd like to guarantee the order of actions. Checks 
```
                 Await state.WaitForAsynchronousOperationsAsync()
                 Await state.AssertNoCompletionSession()
                 Assert.Contains("System", state.GetLineTextFromCaretPosition())
```
should not be started before `task1` completion. `task2` is added here just for consistency.